### PR TITLE
LSR-979 Cookie Login (Android)

### PR DESCRIPTION
**Pull request by victorg1991**

### CODE REVIEW

File: android/library/src/main/java/com/liferay/mobile/screens/auth/login/LoginScreenlet.java [hunk 1]
```diff
@@ -207,7 +210,10 @@ protected View createScreenletView(Context context, AttributeSet attributes) {
 
 	@Override
 	protected BaseLoginInteractor createInteractor(String actionName) {
-		if (BASIC_AUTH.equals(actionName)) {
+		if (authenticationType.equals(AuthenticationType.COOKIE)) {
+			return new LoginCookieInteractor();
+		}
+		else if (BASIC_AUTH.equals(actionName)) {
```
Comment (jmnavarro) [inline android/library/src/main/java/com/liferay/mobile/screens/auth/login/LoginScreenlet.java]: @victorg1991 the action name should be always the same, and we'll rely on "authenticationType" to decide the interactor to be created.
Check iOS as well

Comment (victorg1991) [inline android/library/src/main/java/com/liferay/mobile/screens/auth/login/LoginScreenlet.java]: Fixed 👍

File: android/library/src/main/java/com/liferay/mobile/screens/auth/login/LoginScreenlet.java [hunk 2]
```diff
@@ -219,16 +223,20 @@ protected BaseLoginInteractor createInteractor(String actionName) {
...
+		else {
+			return new LoginBasicInteractor();
+		}
 	}
 
 	@Override
 	protected void onUserAction(String userActionName, BaseLoginInteractor interactor, Object... args) {
-		if (BASIC_AUTH.equals(userActionName)) {
-
+		if (authenticationType.equals(AuthenticationType.COOKIE)) {
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/auth/login/LoginScreenlet.java]: equals should be inverted

File: android/library/src/main/java/com/liferay/mobile/screens/base/interactor/BaseCacheReadInteractor.java [hunk 3]
```diff
@@ -30,45 +34,79 @@ public void start(final Object... args) {
...
+			}
+		}
+	}
+
+	protected Object[] addRetryingVarArg(Object... args) {
+		if (this instanceof BaseListInteractor) {
+			Object[] newArgs = new Object[args.length + 1];
+			newArgs[args.length] = true;
+
+			System.arraycopy(args, 0, newArgs, 0, args.length);
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/base/interactor/BaseCacheReadInteractor.java]: why this?

File: android/library/src/main/java/com/liferay/mobile/screens/base/interactor/BaseInteractor.java [hunk 4]
```diff
@@ -14,6 +16,7 @@
 public abstract class BaseInteractor<L, E extends BasicEvent> implements Interactor<L> {
 
 	protected L listener;
+	protected boolean retried;
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/base/interactor/BaseInteractor.java]: Can you store this in the event?

File: android/library/src/main/java/com/liferay/mobile/screens/context/AuthenticationType.java [hunk 5]
```diff
@@ -19,5 +19,21 @@
...
-	BASIC, OAUTH, VOID
+	BASIC(0), OAUTH(1), COOKIE(2), VOID(3);
+
+	private final int value;
+
+	AuthenticationType(int value) {
+		this.value = value;
+	}
+
+	public static AuthenticationType getValue(int value) {
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/context/AuthenticationType.java]: I think enums can already do this automatically or at least AuthenticationType.values()[value]

File: android/library/src/main/java/com/liferay/mobile/screens/context/LiferayServerContext.java [hunk 6]
```diff
@@ -143,4 +162,30 @@ public Response intercept(Chain chain) throws IOException {
...
+			}
+		}
+
+		return builder.build();
+	}
+
+	public static Map<String, String> getAuthHeaders() {
+		try {
+			com.liferay.mobile.android.http.Request authRequest =
+				new com.liferay.mobile.android.http.Request(null, null, null, null, 0);
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/context/LiferayServerContext.java]: so pretty...

File: android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/imagegallery/DetailImageActivity.java [hunk 7]
```diff
@@ -27,7 +28,7 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
 			ImageEntry imageEntry = (ImageEntry) intent.getExtras().get(GALLERY_SCREENLET_IMAGE_DETAILED);
 
 			ImageView detailedImageView = (ImageView) findViewById(R.id.detailed_image);
-			Picasso.with(this).load(imageEntry.getImageUrl()).into(detailedImageView);
+			PicassoScreens.load(imageEntry.getImageUrl()).into(detailedImageView);
```
Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/viewsets/defaultviews/imagegallery/DetailImageActivity.java]: 👍

File: android/library/src/main/java/com/liferay/mobile/screens/base/interactor/BaseInteractor.java (hunk 4 above)
Comment (victorg1991) [inline android/library/src/main/java/com/liferay/mobile/screens/base/interactor/BaseInteractor.java]: The event is created after the retries :(

Comment (nhpatt) [inline android/library/src/main/java/com/liferay/mobile/screens/base/interactor/BaseInteractor.java]: Ok... then we have a problem with rotation changes, but we have to refactor all of this eitherway

**Comment by nhpatt** [comment]
LGTM

**Pull request merged**